### PR TITLE
[improve][broker] Support get/remove permissions for AuthorizationProvider

### PR DESCRIPTION
**Pull request by Technoboy-**
### Motivation

This is a follow-up improvement for AuthorizationProvider.  As discussed in the [#20478-comment](https://github.com/apache/pulsar/pull/20478#issuecomment-1575333713), we need more methods to avoid admin handling permissions directly.

### Modification
1. Add `removePermissionsAsync`: when deleting partitioned topics, we need to delete the authentication data.
2. Add `getPermissionsAsync`: we have topic level API to get all the permissions. so we need this.
3. Add `getSubscriptionPermissionsAsync`: we have topic level API to get all subscription permissions.
4. Add `getPermissionsAsync`: we have namespace level API to get all the permissions.

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete`

### CODE REVIEW

**Review by michaeljmarshall — COMMENTED** [review]
Nice direction @Technoboy-. I'll review again when it is ready for review.

**Comment by Technoboy-** [comment]
```
Error:  Medium: org.apache.pulsar.websocket.WebSocketService.getAuthorizationService() may expose internal representation by returning WebSocketService.authorizationService [org.apache.pulsar.websocket.WebSocketService] At WebSocketService.java:[line 177] EI_EXPOSE_REP
[INFO] 


To see bug detail using the Spotbugs GUI, use the following command "mvn spotbugs:gui"
```
For the above spotbugs issue, I have added this fix
https://github.com/apache/pulsar/pull/20496/commits/023781c56cf7e83e64da69df41a181dc159e4989

**Comment by Technoboy-** [comment]
> Nice direction @Technoboy-. I'll review again when it is ready for review.

Thanks

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/AuthorizationProvider.java [hunk 1]
```diff
@@ -428,4 +429,48 @@ default Boolean allowTopicPolicyOperation(TopicName topicName,
             throw new RestException(e.getCause());
         }
     }
+
+    /**
+     * Remove authorization-action permissions on a topic to the given client.
```
Comment (michaeljmarshall) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/AuthorizationProvider.java:434]: Nit: this is not for a given client, but for all clients.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/AuthorizationProvider.java [hunk 2]
```diff
@@ -428,4 +429,48 @@ default Boolean allowTopicPolicyOperation(TopicName topicName,
...
+     * @return CompletableFuture<Void>
+     */
+    default CompletableFuture<Void> removePermissionsAsync(TopicName topicName) {
+        return FutureUtil.failedFuture(new IllegalStateException(
+                String.format("removePermissionsAsync on topicName %s is not supported by the Authorization",
+                        topicName)));
+    }
+
+    /**
+     * Get authorization-action permissions on a topic to the given client.
```
Comment (michaeljmarshall) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/AuthorizationProvider.java:445]: Nit: this is the permission for the whole topic, not just for a given client.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/AuthorizationProvider.java [hunk 3]
```diff
@@ -428,4 +429,48 @@ default Boolean allowTopicPolicyOperation(TopicName topicName,
...
+     * @param topicName
+     * @return CompletableFuture<Map<String, Set<AuthAction>>>
+     */
+    default CompletableFuture<Map<String, Set<AuthAction>>> getPermissionsAsync(TopicName topicName) {
```
Comment (michaeljmarshall) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/AuthorizationProvider.java:449]: One interesting way we could consider expanding this method is to add the role of the user requesting this operation. As it is currently implemented, the broker will first verify that the role is a tenant admin or a superuser and then will call this method. It seems simpler to defer that kind of checking to the authentication provider. However, we don't do that kind of thing for any of the other methods, so that might not be consistent with the current design. What do you think @Technoboy-?

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/AuthorizationProvider.java [hunk 4]
```diff
@@ -428,4 +429,48 @@ default Boolean allowTopicPolicyOperation(TopicName topicName,
...
+        return FutureUtil.failedFuture(new IllegalStateException(
+                String.format("getPermissionsAsync on topicName %s is not supported by the Authorization",
+                        topicName)));
+    }
+
+    /**
+     * Get authorization-action permissions on a topic to the given client.
```
Comment (michaeljmarshall) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/AuthorizationProvider.java:456]: Nit: it's for all clients.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/AuthorizationProvider.java [hunk 5]
```diff
@@ -428,4 +429,48 @@ default Boolean allowTopicPolicyOperation(TopicName topicName,
...
+     * @return CompletableFuture<Map<String, Set<String>>>
+     */
+    default CompletableFuture<Map<String, Set<String>>> getSubscriptionPermissionsAsync(NamespaceName namespaceName) {
+        return FutureUtil.failedFuture(new IllegalStateException(
+                String.format("getSubscriptionPermissionsAsync on namespace %s is not supported by the Authorization",
+                        namespaceName)));
+    }
+
+    /**
+     * Get authorization-action permissions on a namespace to the given client.
```
Comment (michaeljmarshall) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/AuthorizationProvider.java:467]: Nit: for all clients.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 6]
```diff
@@ -843,16 +813,9 @@ private CompletableFuture<Void> internalRemovePartitionsTopicNoAutocreationDisab
                 }).collect(Collectors.toList()));
     }
 
-    private CompletableFuture<Void> internalRemovePartitionsAuthenticationPoliciesAsync(int numPartitions) {
+    private CompletableFuture<Void> internalRemovePartitionsAuthenticationPoliciesAsync() {
```
Comment (michaeljmarshall) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:816]: What is the motivation for removing the partition logic here? I don't think we can do that here, but I might be missing some context.

My understanding is that we need to remove permission for the parent and all of its sub partitions.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/buffer/TransactionBufferCloseTest.java [hunk 7]
```diff
@@ -61,7 +61,7 @@ public Object[][] isPartition() {
         };
     }
 
-    @Test(timeOut = 10_000, dataProvider = "isPartition")
+    @Test(dataProvider = "isPartition")
```
Comment (michaeljmarshall) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/buffer/TransactionBufferCloseTest.java:64]: This seems like it belongs in a separate PR.

**Review by michaeljmarshall — COMMENTED** [review]
The migration code looks great. I left several comments.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java (hunk 6 above)
Comment (Technoboy-) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:816]: yes, you're right.  We will `remove permission` in the BrokerService#deleteTopic. So we just need to remove the topic without partitioned here.  
See line 848, I have deleted them together.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/buffer/TransactionBufferCloseTest.java (hunk 7 above)
Comment (Technoboy-) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/buffer/TransactionBufferCloseTest.java:64]: rollbacked

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/AuthorizationProvider.java (hunk 3 above)
Comment (Technoboy-) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/AuthorizationProvider.java:449]: Currently, my design and implementation are simple. It's only for Admin API usage.
For this `getPermissionsAsync`, it's only used for `getPermissionsOnTopic` in Admin
If we change `getPermissionsAsync(TopicName topicName, String role)`, I think it's really match other methods in the AuthorizationProvider, but we don't need it right now.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/AuthorizationProvider.java [hunk 8]
```diff
@@ -428,4 +429,48 @@ default Boolean allowTopicPolicyOperation(TopicName topicName,
...
+                String.format("getPermissionsAsync on topicName %s is not supported by the Authorization",
+                        topicName)));
+    }
+
+    /**
+     * Get authorization-action permissions on a topic.
+     * @param namespaceName
+     * @return CompletableFuture<Map<String, Set<String>>>
+     */
+    default CompletableFuture<Map<String, Set<String>>> getSubscriptionPermissionsAsync(NamespaceName namespaceName) {
```
Comment (mattisonchao) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/AuthorizationProvider.java:460]: Should we support deleting it?

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/PulsarAuthorizationProvider.java [hunk 9]
```diff
@@ -641,4 +643,131 @@ public CompletableFuture<Boolean> validateTenantAdminAccess(String tenantName, S
...
+            if (readonly) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Policies are read-only. Broker cannot do read-write operations");
+                }
+                throw new IllegalStateException("policies are in readonly mode");
+            }
+            return pulsarResources.getNamespaceResources().getPoliciesAsync(topicName.getNamespaceObject())
+                    .thenCompose(policies -> {
+                        if (!policies.isPresent()
+                                || !policies.get().auth_policies.getTopicAuthentication()
```
Comment (mattisonchao) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/PulsarAuthorizationProvider.java:658]: Does `auth_policies` can be null?

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/PulsarAuthorizationProvider.java [hunk 10]
```diff
@@ -641,4 +643,131 @@ public CompletableFuture<Boolean> validateTenantAdminAccess(String tenantName, S
                 });
     }
 
+    @Override
+    public CompletableFuture<Void> removePermissionsAsync(TopicName topicName) {
```
Comment (mattisonchao) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/PulsarAuthorizationProvider.java:647]: Could you please add an NPE check for `topicName` since it is a public API?

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/PulsarAuthorizationProvider.java [hunk 11]
```diff
@@ -641,4 +643,131 @@ public CompletableFuture<Boolean> validateTenantAdminAccess(String tenantName, S
...
+            return pulsarResources.getNamespaceResources().getPoliciesAsync(topicName.getNamespaceObject())
+                    .thenCompose(policies -> {
+                        if (!policies.isPresent()
+                                || !policies.get().auth_policies.getTopicAuthentication()
+                                .containsKey(topicName.toString())) {
+                            return CompletableFuture.completedFuture(null);
+                        }
+                        return pulsarResources.getNamespaceResources().
+                                setPoliciesAsync(topicName.getNamespaceObject(), policies2 -> {
+                                    policies2.auth_policies.getTopicAuthentication().remove(topicName.toString());
```
Comment (mattisonchao) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/PulsarAuthorizationProvider.java:664]: Should we add `if (!policies.isPresent()) {`?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java (hunk 6 above)
Comment (mattisonchao) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:816]: I saw `removePermissionsAsync` only deleted the parent info but no sub info in `internalRemovePartitionsAuthenticationPoliciesAsync` method. 
Do you mean the rest of the sub info will delete by `BrokerService#deleteTopic`?

**Review by mattisonchao — APPROVED** [review]

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/PulsarAuthorizationProvider.java (hunk 9 above)
Comment (Technoboy-) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/PulsarAuthorizationProvider.java:658]: It won't, see https://github.com/apache/pulsar/blob/4586852f30902b5247d907ce4cc9360152db1e21/pulsar-client-admin-api/src/main/java/org/apache/pulsar/common/policies/data/Policies.java#L33-L37

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/PulsarAuthorizationProvider.java (hunk 10 above)
Comment (Technoboy-) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/PulsarAuthorizationProvider.java:647]: Maybe a bit redundant

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/PulsarAuthorizationProvider.java (hunk 11 above)
Comment (Technoboy-) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/PulsarAuthorizationProvider.java:664]: See line-657

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java (hunk 6 above)
Comment (Technoboy-) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:816]: yes, right, it's duplicated here.
